src/network/network_gui.cpp

```diff
@@ -1806,166 +1793,543 @@ static void PopupClientList(ClientID client_id, int x, int y)
+			case WID_CL_MATRIX: {
+				Dimension d = GetSpriteSize(SPR_COMPANY_ICON);
+				this->line_height = std::max(d.height, (uint)FONT_HEIGHT_NORMAL) + WD_MATRIX_TOP + WD_MATRIX_BOTTOM;
 
-		uint width = 100; // Default width
-		for (const NetworkClientInfo *ci : NetworkClientInfo::Iterate()) {
-			width = std::max(width, GetStringBoundingBox(ci->client_name).width);
+				resize->width = 1;
+				resize->height = this->line_height;
+				fill->height = this->line_height;
+				size->height = std::max(size->height, 5 * this->line_height);
+				break;
+			}
 		}
+	}
+
+	void OnResize() override
+	{
+		this->vscroll->SetCapacityFromWidget(this, WID_CL_MATRIX);
+	}
 
-		size->width = WD_FRAMERECT_LEFT + this->server_client_width + this->icon_size.width + WD_FRAMERECT_LEFT + width + WD_FRAMERECT_RIGHT;
+	void SetStringParameters(int widget) const override
+	{
+		switch (widget) {
+			case WID_CL_SERVER_VISIBILITY:
+				SetDParam(0, _server_visibility_dropdown[_settings_client.network.server_advertise]);
+				break;
+		}
 	}
 
-	void OnPaint() override
+	void OnClick(Point pt, int widget, int click_count) override
 	{
-		/* Check if we need to reset the height */
-		if (!this->CheckClientListHeight()) return;
+		switch (widget) {
+			case WID_CL_SERVER_NAME_EDIT:
+				if (!_network_server) break;
 
-		this->DrawWidgets();
+				this->query_widget = WID_CL_SERVER_NAME_EDIT;
+				SetDParamStr(0, _settings_client.network.server_name);
+				ShowQueryString(STR_JUST_RAW_STRING, STR_NETWORK_CLIENT_LIST_SERVER_NAME_QUERY_CAPTION, NETWORK_NAME_LENGTH, this, CS_ALPHANUMERAL, QSF_ENABLE_DEFAULT | QSF_LEN_IN_CHARS);
+				break;
+
+			case WID_CL_CLIENT_NAME_EDIT:
+				this->query_widget = WID_CL_CLIENT_NAME_EDIT;
+				SetDParamStr(0, _settings_client.network.client_name);
+				ShowQueryString(STR_JUST_RAW_STRING, STR_NETWORK_CLIENT_LIST_PLAYER_NAME_QUERY_CAPTION, NETWORK_CLIENT_NAME_LENGTH, this, CS_ALPHANUMERAL, QSF_ENABLE_DEFAULT | QSF_LEN_IN_CHARS);
+				break;
+
+			case WID_CL_SERVER_VISIBILITY:
+				if (!_network_server) break;
+
+				ShowDropDownMenu(this, _server_visibility_dropdown, _settings_client.network.server_advertise, WID_CL_SERVER_VISIBILITY, 0, 0);
+				break;
+
+			case WID_CL_MATRIX:
+				uint index = this->vscroll->GetScrolledRowFromWidget(pt.y, this, WID_CL_MATRIX);
+				NWidgetBase *widget = this->GetWidget<NWidgetBase>(WID_CL_MATRIX);
+
+				bool rtl = _current_text_dir == TD_RTL;
+				uint x = rtl ? (uint)WD_FRAMERECT_LEFT : widget->current_x - WD_FRAMERECT_RIGHT;
+
+				/* Find the buttons for this row. */
+				auto button_find = this->buttons.find(index);
+				if (button_find == this->buttons.end()) break;
+
+				/* Check if we clicked on any of the buttons. */
+				for (auto &button : button_find->second) {
+					uint left = rtl ? x : x - button.width;
+					uint right = rtl ? x + button.width : x;
+
+					if (IsInsideMM(pt.x, left, right)) {
+						if (button.company_proc != nullptr) {
```

Comment:
I would put these 6 lines in the Button class in something like a OnClick(NCLW *, Point). Then you do not need to know any of the internals of the button here.